Midi 15/composer recognition

WojciechMat opened the pull request:
Draft pull request created as soon as possible,
will add description tommorow (after cleaning the code)

WojciechMat committed:
preparing data for training

WojciechMat committed:
first chaotic implementation of not-working model

WojciechMat committed:
docstrings and less chaos

WojciechMat committed:
cosmetic changes

WojciechMat committed:
cosmetic changes

WojciechMat commented:
Very simple, sequential neural network has been implemented to learn to classify lists of 128 note pitches into two classes:
Frédéric Chopin" and "Ludwig van Beethoven".
I chose these composers due to the similar total number of notes in maestro dataset.
### Number of samples:
**Train split:**
![image](https://github.com/Nospoko/bert-midi-evals/assets/74838859/be580ca8-29d0-4a9f-9bb3-ff6eeb57795a)
**Test split:**
![image](https://github.com/Nospoko/bert-midi-evals/assets/74838859/76af591a-d9e2-4b4b-b8cb-35485d412cfd)
### Accuracy:
The results seem to be pretty random, probably due to the fact that getting pitches alone from the data is not really a great feature extraction method. The simplicity of NN is also not helping.

![image](https://github.com/Nospoko/bert-midi-evals/assets/74838859/2b7c7116-9883-472a-926d-4a0652d9c02d)

I propose to experiment with simple time-data quantization and creation of a more complex model in the next step.

roszcz commented on src/model1.pth:
We shouldn't commit binary files to repositories :)

roszcz commented on src/main.py:
```diff
@@ -0,0 +1,130 @@
+from torch.utils.data import DataLoader
+
+from model import PitchSeqNN
+from src.data.dataset import ComposerClassificationDataset
```
```suggestion
from model import PitchSeqNN
from data.dataset import ComposerClassificationDataset
```

This may require some changes to your module design, but I think it's better to not reference the `src` directory in code. Personally, would I would get rid of it completely, and `main.py`, `model.py` and `data/` could be root level.

roszcz commented on src/main.py:
```diff
@@ -0,0 +1,130 @@
+from src.data.dataset import ComposerClassificationDataset
+
+BATCH_SIZE = 16
+N_EPOCHS = 100
```
Don't worry about it for this PR, but please take a look at https://hydra.cc/docs/intro/ - it's a tool I've been recently using to manage configs and I think it's great!

roszcz commented on src/main.py:
```diff
@@ -0,0 +1,130 @@
+    df["composer"] = df["composer"].apply(lambda x: 0 if x == dataset.selected_composers[0] else 1)
+    # take only ['pitch'] column as an input tensor, normalize it to belong to (0, 1]
+    df["notes"] = df["notes"].apply(lambda x: x["pitch"])
+    df["notes"] = df["notes"].apply(divide)
```
```suggestion
    df["pitches"] = df["notes"].apply(lambda x: x["pitch"])
    df["normalized_pitches"] = df["notes"].apply(divide)
```

I think overwriting existing columns can quickly lead to confusion. I would also suggest looking for a different way to do the normalization, maybe even after the tensor is created from the dataframe, rather than inside of it.

WojciechMat committed:
requirements

roszcz commented:
> The results seem to be pretty random, probably due to the fact that getting pitches alone from the data is not really a great feature extraction method. The simplicity of NN is also not helping.

I wouldn't give up on this yet - my first guess would be that your model is too deep for this problem, can you try decreasing the number of layers, to see if you can get past the 50% mark?

Second thing to check would be the loss function, just to be sure that your feeding it with the properly scaled data and that nothing is exploding in there :)

WojciechMat commented:
Okay, I actually tired another model, with less epochs so without overfitting it to the train data, and got these results:

![image](https://github.com/Nospoko/bert-midi-evals/assets/74838859/e34e32d5-0a42-421b-9a50-ce89a6dad92b)

I will continue the experiments.

roszcz commented on src/main.py:
```diff
@@ -0,0 +1,130 @@
+    return model
+
+
+def test_model(model, path: Optional[str] = None):
```
```suggestion
def test_model(model):
```

I think the model testing logic should be separated from checkpoint loading, i.e.:

```python
model = train()
test_model(model)

model = PitchSeqNN()
model.load_state_dict(...)
test_model(model)
```

WojciechMat replied:
Yes! I actually already changed that just have not commited yet - sorry

roszcz commented on src/main.py:
```diff
@@ -0,0 +1,130 @@
+    samples = [(torch.tensor(row["notes"], dtype=torch.float32), torch.tensor(row["composer"])) for _, row in df.iterrows()]
+    while len(samples) % BATCH_SIZE != 0:
+        samples.pop(-1)
+    dataset.samples = samples
```
```suggestion
    return samples
```

I'm not a fan of functions that modify objects secretly - I think it's much easier to understand something like this:

```python
dataset.samples = modify_samples(dataset)
```

Here, there's no need to read `modify_samples` to know what's happening.

WojciechMat replied:
Okay! Thanks

roszcz commented on src/main.py:
```diff
@@ -0,0 +1,130 @@
+    df["notes"] = df["notes"].apply(divide)
+    print(df.groupby("composer").size())
+    samples = [(torch.tensor(row["notes"], dtype=torch.float32), torch.tensor(row["composer"])) for _, row in df.iterrows()]
+    while len(samples) % BATCH_SIZE != 0:
+        samples.pop(-1)
```
```suggestion
```

We should be able to deal with number of records that does not split into an even number of batches - I think that torch loaders support this by default.

WojciechMat replied:
When I did not include this line, I got errors regarding incorrect shapes, but I’ll look into that.

roszcz commented on src/main.py:
```diff
@@ -0,0 +1,130 @@
+    df["notes"] = df["notes"].apply(divide)
+    print(df.groupby("composer").size())
+    samples = [(torch.tensor(row["notes"], dtype=torch.float32), torch.tensor(row["composer"])) for _, row in df.iterrows()]
+    while len(samples) % BATCH_SIZE != 0:
+        samples.pop(-1)
```
```suggestion
```

We should be able to deal with number of records that does not split into an even number of batches - I think that torch loaders support this by default.

roszcz commented on src/main.py:
```diff
@@ -0,0 +1,130 @@
+
+            running_loss += loss.item()
+            # Periodically print the running loss for monitoring the training progress.
+            if i % 500 == 499:
+                print(f"[{epoch + 1}, {i + 1:5d}] loss: {running_loss / 500:.3f}")
+                running_loss = 0.0
```
```suggestion
      running_loss = running_loss / len(train_dataloader)                                                                    
      val_loss = test_model(model)
      print(f"[{epoch + 1}] loss: {running_loss:.3f} val loss: {val_loss:.3f}")  
      running_loss = 0.0

```

This is a train loop monitoring method that I think would work better here. This suggestion is more of a pseudo-code, but here's the gist of what's the point:

1. Dataset and model are both small, so we can check metrics every epoch. (I'm not sure if it's clear that indent in my suggestion is supposed to be on the epoch level)
2. If possible, it's best to monitor the same metric on both splits, so the print every epoch could look like this `[epoch 76] training loss: 0.021, validation loss: 0.7`. You could also include accuracy here.

WojciechMat replied:
Thank you!!

roszcz commented:
> Okay, I actually tired another model, with less epochs so without overfitting it to the train data, and got these results:
> 
> ![image](https://user-images.githubusercontent.com/74838859/258424010-e34e32d5-0a42-421b-9a50-ce89a6dad92b.png)
> 
> I will continue the experiments.

To me this still looks like a model guessing randomly (by tossing an unbalanced coin :))

Did you actually observe overfitting with this? When I'm running this locally, the loss value stays at ~0.7 and it looks like the model is not learning/memorizing. Overfitting is only a tragedy if you don't know that it's happening, otherwise it can be used as a diagnostic tool - here's a great blog post that touches upon this topic: http://karpathy.github.io/2019/04/25/recipe/

One change to try, would be to modify model inputs in a way that makes them mean 0:

```python
def divide(lst):                                                                                                           
    return [(num - 64) / 127.0 for num in lst] 
```

WojciechMat commented:
I have observed that it remembers some pieces by testing on a train set, with ~60% accuracy compared to 49% with test dataset. I will update this branch soon.

WojciechMat committed:
second model, updated train loop output

WojciechMat commented:
Here are results of overfitted model on train data (not test, like the msg suggests):

![image](https://github.com/Nospoko/bert-midi-evals/assets/74838859/8bae8c39-2e81-40ed-91dc-ee0e26742412)

I will add more functionality, experiment with more models and try to make the model actually working, not just remembering, tomorrow.
It has ~51% accuracy on test data, so still random.

WojciechMat committed:
better project architecture, testing some simple NNs

WojciechMat committed:
cosmetic changes

WojciechMat committed:
accuracy metric during training

WojciechMat committed:
requirements update

WojciechMat committed:
plot model history

WojciechMat commented:
### Metrics
I have added accuracy, val_loss and val_accuracy monitoring during training:

![image](https://github.com/Nospoko/bert-midi-evals/assets/74838859/78c262e3-42ae-473e-ad0c-1e393416d367) 

I have also added a function to plot the loss and accuracy curves.

### Running tests
I have added some really simple NNs and tested them on samples of 64 pitches per label, in batches of 16. Here are the train histories:

**PitchSeqNN:**
![loss1](https://github.com/Nospoko/bert-midi-evals/assets/74838859/d92cb0c4-a900-4660-b636-4a9dbd71c37e)
![accuracy1](https://github.com/Nospoko/bert-midi-evals/assets/74838859/58ae51b9-4489-46bd-9ae9-0116a425affa)

**PitchSeqNNv2:**
![loss2](https://github.com/Nospoko/bert-midi-evals/assets/74838859/d69e7565-9259-49cb-a6b2-572ce6d6ebf5)
![accuracy2](https://github.com/Nospoko/bert-midi-evals/assets/74838859/8e2cc3d5-e2d3-4d66-9b5b-1af2a6329f61)

**PitchSeqNNv3:**
The same.
**PitchSeqNNv4:**
The same.

roszcz commented on main.py:
```diff
@@ -1,53 +1,223 @@
+import torch.nn as nn
+from torch.utils.data import DataLoader
+
+from utils import plot_loss_curves
```
```
ModuleNotFoundError: No module named 'utils'
```

Missing file? :)

WojciechMat replied:
yess, sorry :))

roszcz commented on main.py:
```diff
@@ -1,53 +1,223 @@
+
+
+def main():
+    models = [PitchSeqNN(), PitchSeqNNv2_64(), PitchSeqNNv3_64(), PitchSeqNNv4_64()]
```
```suggestion
    layers = [
      [64, 128, 256, 64, 2],
      [64, 128, 64, 2],
      ...
    ]
    for layer in layers:
      model = PitchNN(layers=layers)
```

As far as I can tell all versions of this model differ only in the number and shape of linear layers. I think it's a better approach to have a single model class and treat those values as parameters, especially if you want to check results for multiple configurations like this.

roszcz commented on main.py:
```diff
@@ -1,53 +1,223 @@
+    val_accuracy = correct / total
+    return val_loss, val_accuracy
+
+
+def test_model(model: nn.Module, test_data: DataLoader, path: Optional[str] = None, classnames=None):
```
```suggestion

def test_model(model: nn.Module, test_data: DataLoader):
```

I'd remove the `path` argument to separate the testing logic from the checkpoint loading logic:

```python
# Prefer this:
model = load_model()
scores = test_model(model)

# Over this:
scores = test_model(model, path_to_load_model)
```

For the `classnames` - I'd say this is a property of the dataset, so you shouldn't need a separate argument for that, and rather get it from the data object your passing:

```python
classnames = dataloader.dataset.classnames
```

roszcz commented on main.py:
```diff
@@ -1,53 +1,223 @@
+    if path is not None:
+        model.load_state_dict(torch.load(path))
+    # containers for counting prediction
+    correct_pred = {classname: 0 for classname in classnames}
+    total_pred = {classname: 0 for classname in classnames}
+    correct = 0
+    total = 0
```
```suggestion
    correct_pred = {classname: 0 for classname in classnames}
```

I think there is no need to count the "total" numbers, you can get this by doing `len(dataloader.dataset)`. One container should be enough to count the correct results as well :)

roszcz commented on main.py:
```diff
@@ -1,53 +1,223 @@
+            out = model(notes)
+            loss = nn.CrossEntropyLoss()(out, labels)
+            print(loss)
+            _, preds = torch.max(out, 1)
```
```suggestion
            preds = out.argmax(1)
```

Argmax is more popular for this in torch, and straightforward to read

roszcz commented on main.py:
```diff
@@ -1,53 +1,223 @@
+            correct += (preds == labels).sum().item()
+            total += labels.size(0)
+
+            for label, pred in zip(labels, preds):
+                if label == pred:
+                    correct_pred[classnames[label]] += 1
+                total_pred[classnames[label]] += 1
+    for classname, correct_count in correct_pred.items():
+        accuracy = 100 * float(correct_count) / total_pred[classname]
+        print(f"Accuracy for class: {classname:5s} is {accuracy:.1f} %")
+    print(f"Accuracy of the network on the test data: {(100 * correct / total):.3f} %")
+    print(f"correctly predicted : {correct_pred}")
```
I think using a [confusion matrix](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.confusion_matrix.html) here would work better. Your version could get problematic to maintain if we were to start adding classes.

roszcz commented on main.py:
```diff
@@ -1,53 +1,223 @@
+            loss = criterion(out, labels)
+            loss_sum += loss.item()
+            _, preds = torch.max(out, 1)
+            for pred, label in zip(preds, labels):
+                if pred == label:
+                    correct += 1
```
```suggestion
            correct += (pred == label).sum()
```

It's almost always better to not have fors and ifs :)

roszcz reviewed (commented):
One generic readability suggestion I have - you could try adding more empty lines between your code - this can add paragraph like structure and improve reader comprehension:

To me, this looks very dense:

```python
def main():
    models = [PitchSeqNN(), PitchSeqNNv2_64(), PitchSeqNNv3_64(), PitchSeqNNv4_64()]
    losses = []
    train_dataloader = get_dataloader(split="train", sequence_length=64, batch_size=16)
    validation_dataloader = get_dataloader(split="validation", sequence_length=64, batch_size=16)
    test_dataloader = get_dataloader(split="test", sequence_length=64, batch_size=16)
    # classnames = ComposerClassificationDataset().selected_composers
    for model in models:
        model, history = train_model(model, train_data=train_dataloader, lr=3e-4, val_data=validation_dataloader)
        losses.append(evaluate(model, test_dataloader))
        plot_loss_curves(history)
    print(losses)
    torch.save(models[np.argmax(losses)].state_dict(), "best.pth")
```

And this is easier to parse visually:

```python
def main():
    # Versions to check
    models = [PitchSeqNN(), PitchSeqNNv2_64(), PitchSeqNNv3_64(), PitchSeqNNv4_64()]

    # Data to use
    train_dataloader = get_dataloader(split="train", sequence_length=64, batch_size=16)
    validation_dataloader = get_dataloader(split="validation", sequence_length=64, batch_size=16)
    test_dataloader = get_dataloader(split="test", sequence_length=64, batch_size=16)

    # Check each version
    losses = []
    for model in models:
        model, history = train_model(model, train_data=train_dataloader, lr=3e-4, val_data=validation_dataloader)
        losses.append(evaluate(model, test_dataloader))
        plot_loss_curves(history)

    print(losses)
    torch.save(models[np.argmax(losses)].state_dict(), "best.pth")
```

I also added some comments in this, mostly quite redundant, but I'd argue that the readability gain is worth it :)

WojciechMat committed:
utils file

WojciechMat committed:
confusion matrix, one changable PitchSeqNN class

WojciechMat commented:
I have implemented requested changes.
### Models:
PitchSeqNN is  now requiring a layers argument:  list specifying the number of neurons in each layer of the neural network. The first element should represent the input size, the last element represents the output size, and the intermediate elements represent the hidden layer sizes.

### Plots:
I have added a function for creating confusion matrixes for any number of classes. The test_model function now plots a confusion matrix instead of counting and printing results.
For example, here are the plots for one of the models:

![loss3](https://github.com/Nospoko/bert-midi-evals/assets/74838859/da8851b5-cdfc-455c-8cfe-481f10df005d)
![accuracy3](https://github.com/Nospoko/bert-midi-evals/assets/74838859/3fa6f878-4a0b-43a1-baaa-8c7521bea005)
![confusion_matrix](https://github.com/Nospoko/bert-midi-evals/assets/74838859/87536272-db5f-43a2-922f-bdb4c16f1b82)

WojciechMat committed:
renaming, moved testing function to utils file
